// @alifd/mobile-layout/lib/common/constant.js
"use strict";

exports.__esModule = true;
exports.VER_ALIGN_ALIAS_MAP = exports.PARAGRAPH_TYPE_ALIAS_MAP = exports.DEFAULT_BREAK_POINTS = exports.ALIGN_ALIAS_MAP = void 0;

/**
 * 默认断点列信息
 */
var DEFAULT_BREAK_POINTS = [{
  width: 200,
  maxContentWidth: 200,
  numberOfColumns: 1
}, {
  width: 320,
  maxContentWidth: 320,
  numberOfColumns: 2
}, {
  width: 750,
  maxContentWidth: 750,
  numberOfColumns: 4
}, {
  width: 1125,
  maxContentWidth: 1125,
  numberOfColumns: 8
}, {
  width: Infinity,
  maxContentWidth: Infinity,
  numberOfColumns: 12
}]; // css 别名处理

exports.DEFAULT_BREAK_POINTS = DEFAULT_BREAK_POINTS;
var ALIGN_ALIAS_MAP = {
  left: 'start',
  right: 'end',
  middle: 'center'
};
exports.ALIGN_ALIAS_MAP = ALIGN_ALIAS_MAP;
var VER_ALIGN_ALIAS_MAP = {
  top: 'start',
  middle: 'center',
  bottom: 'end'
};
exports.VER_ALIGN_ALIAS_MAP = VER_ALIGN_ALIAS_MAP;
var PARAGRAPH_TYPE_ALIAS_MAP = {
  body1: 'body-1',
  body2: 'body-2',
  display1: 'display-1',
  display2: 'display-2',
  display3: 'display-3'
};
exports.PARAGRAPH_TYPE_ALIAS_MAP = PARAGRAPH_TYPE_ALIAS_MAP;
