// @alifd/mobile-layout/es/local/col-flex.js
var _excluded = ["children", "width", "height", "className", "align", "gap", "style"],
    _excluded2 = ["height", "style", "autoFit"],
    _excluded3 = ["height"];

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import { createElement, useContext } from 'rax';
import classNames from 'classnames';
import Children from 'rax-children';
import isValidElement from 'rax-is-valid-element';
import cloneElement from 'rax-clone-element';
import Context from '../common/context';
import { ALIGN_ALIAS_MAP } from '../common/constant';
import { isValidGap, getGapVal, wrapUnit } from '../utils';

var ColFlex = function (props) {
  var children = props.children,
      width = props.width,
      height = props.height,
      className = props.className,
      align = props.align,
      gapProp = props.gap,
      style = props.style,
      others = _objectWithoutPropertiesLoose(props, _excluded);

  var _useContext = useContext(Context),
      prefix = _useContext.prefix,
      gridGap = _useContext.gridGap;

  var gap = getGapVal(gridGap, gapProp);
  var newChildren = Children.map(children, function (child, index) {
    if (isValidElement(child)) {
      var _child$props = child == null ? void 0 : child.props,
          childHeight = _child$props.height,
          childStyle = _child$props.style,
          childAutoFit = _child$props.autoFit,
          otherChildProps = _objectWithoutPropertiesLoose(_child$props, _excluded2);

      var _ref = childStyle || {},
          childStyleHeight = _ref.height,
          otherChildStyle = _objectWithoutPropertiesLoose(_ref, _excluded3);

      var flex;
      var validHeight;

      if (childHeight && childHeight !== '') {
        validHeight = childHeight;
      }

      if (childStyleHeight && childStyleHeight !== '') {
        validHeight = childStyle.height;
      }

      if (childAutoFit || validHeight) {
        flex = "0 0 auto";
      } else {
        flex = '1 1 0';
      }

      return cloneElement(child, _extends({}, otherChildProps, {
        key: "layout-col-flex-" + index,
        height: validHeight,
        style: _extends({
          flex: flex
        }, otherChildStyle)
      }));
    }

    return child;
  });

  var newStyle = _extends({
    alignItems: ALIGN_ALIAS_MAP[align] || align,
    justifyContent: 'stretch'
  }, width && width > 0 ? {
    width: wrapUnit(width)
  } : null, height && height > 0 ? {
    height: wrapUnit(height)
  } : null, isValidGap(gap) && gap !== 'auto' ? {
    gap: gap
  } : null, style);

  return createElement("div", _extends({}, others, {
    className: classNames(className, prefix + "col-flex"),
    style: newStyle
  }), newChildren);
};

ColFlex.displayName = 'ColFlex';
export default ColFlex;
