// @alifd/mobile-layout/es/local/grid.js
var _excluded = ["className", "children", "style", "align", "verAlign", "renderItem", "rows", "cols", "rowGap", "colGap", "minWidth", "maxWidth"];

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import { createElement, useContext } from 'rax';
import classNames from 'classnames';
import Context from '../common/context';
import View from '../common/view';
import { getGapVal, isValidGap, wrapUnit } from '../utils';

/**
 * 网格布局
 * @param props
 * @constructor
 */
var Grid = function (props) {
  var _classNames;

  var className = props.className,
      children = props.children,
      style = props.style,
      alignProp = props.align,
      verAlignProp = props.verAlign,
      renderItem = props.renderItem,
      rows = props.rows,
      cols = props.cols,
      rowGapProp = props.rowGap,
      colGapProp = props.colGap,
      minWidth = props.minWidth,
      maxWidth = props.maxWidth,
      others = _objectWithoutPropertiesLoose(props, _excluded);

  var _useContext = useContext(Context),
      prefix = _useContext.prefix,
      gridGap = _useContext.gridGap; // 行列 gap 判断顺序：rowGap > Context.gridGap > default:'medium'


  var rowGap = getGapVal(gridGap, rowGapProp);
  var colGap = getGapVal(gridGap, colGapProp);
  var gridTemplateColumns = "repeat(" + cols + ", 1fr)";

  if (minWidth && maxWidth) {
    gridTemplateColumns = "repeat(auto-fill, minmax(" + wrapUnit(minWidth) + ", " + wrapUnit(maxWidth) + "))";
  } else if (minWidth && !maxWidth) {
    gridTemplateColumns = "repeat(auto-fit, minmax(" + wrapUnit(minWidth) + ", auto))";
  } else if (!minWidth && maxWidth) {
    gridTemplateColumns = "repeat(auto-fill, minmax(auto, " + wrapUnit(maxWidth) + "))";
  }

  var newStyle = _extends({
    display: 'grid',
    gridTemplateColumns: gridTemplateColumns,
    gridTemplateRows: "repeat(" + rows + ", 1fr)"
  }, isValidGap(rowGap) && rowGap !== 'auto' ? {
    gridRowGap: rowGap
  } : null, isValidGap(colGap) && colGap !== 'auto' ? {
    gridColumnGap: colGap
  } : null, style); // notice: 小程序渲染的特殊性， justifyItems 和 alignItems 无法在 style 中写入


  var align = alignProp;
  var valign = verAlignProp;
  return createElement(View, _extends({}, others, {
    className: classNames(className, prefix + "grid", (_classNames = {}, _classNames[prefix + "grid-align--" + align] = align, _classNames[prefix + "grid-valign--" + valign] = valign, _classNames)),
    style: newStyle
  }), renderItem ? function renderChildren() {
    return Array.from(new Array(rows)).map(function (_, row) {
      return Array.from(new Array(cols)).map(function (__, col) {
        return renderItem(row, col);
      });
    });
  }() : children);
};

Grid.displayName = 'Grid';
Grid._typeMark = 'Grid';
Grid.defaultProps = {
  spacing: 'medium',
  rows: 1,
  cols: 1,
  verAlign: 'stretch',
  align: 'stretch'
};
export default Grid;
