// @alifd/mobile-layout/es/page/cell.js
var _excluded = ["className", "children", "verAlign", "width", "height", "block", "direction", "gap", "align", "style"];

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import { createElement, useContext } from 'rax';
import classNames from 'classnames';
import View from '../common/view';
import { ALIGN_ALIAS_MAP, VER_ALIGN_ALIAS_MAP } from '../common/constant';
import Context from '../common/context';
import { isValidGap, wrapUnit } from '../utils';

/**
 * 单元格容器(默认垂直方向的 flex 布局容器)
 */
var Cell = function (props) {
  var _classNames;

  var className = props.className,
      children = props.children,
      verAlign = props.verAlign,
      width = props.width,
      height = props.height,
      block = props.block,
      direction = props.direction,
      gap = props.gap,
      align = props.align,
      style = props.style,
      others = _objectWithoutPropertiesLoose(props, _excluded);

  var _useContext = useContext(Context),
      prefix = _useContext.prefix;

  var newStyle = _extends({}, !block ? _extends({
    display: 'flex',
    flexDirection: direction === 'ver' ? 'column' : 'row'
  }, verAlign ? {
    justifyContent: VER_ALIGN_ALIAS_MAP[verAlign] || verAlign
  } : null) : null, {
    width: width ? wrapUnit(width) : 'auto'
  }, isValidGap(gap) ? {
    display: 'grid',
    width: '100%',
    justifyContent: 'none',
    justifyItems: ALIGN_ALIAS_MAP[align] || align,
    gridGap: wrapUnit(gap)
  } : null, height ? {
    height: wrapUnit(height)
  } : null, style);

  return createElement(View, _extends({}, others, {
    className: classNames(prefix + "cell", className, (_classNames = {}, _classNames[prefix + "cell-align--" + align] = align, _classNames)),
    style: newStyle
  }), children);
};

Cell.displayName = 'Cell';
Cell._typeMark = 'Cell';
Cell.defaultProps = {
  direction: 'ver',
  verAlign: 'top',
  block: false
};
export default Cell;
