// @alifd/mobile-layout/es/page/footer.js
var _excluded = ["children", "className", "divider", "noTopPadding", "fixed", "mode", "containerClassName", "containerStyle", "style"];

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import { createElement, useContext } from 'rax';
import classNames from 'classnames';
import View from '../common/view';
import Context from '../common/context';
import { isValidGap, wrapUnit } from '../utils';
import SafeArea from '../common/safe-area';

/**
 * 终章
 * @param props
 * @constructor
 */
var Footer = function (props) {
  var _classNames;

  var children = props.children,
      className = props.className,
      divider = props.divider,
      noTopPadding = props.noTopPadding,
      fixed = props.fixed,
      mode = props.mode,
      containerClassName = props.containerClassName,
      containerStyle = props.containerStyle,
      style = props.style,
      others = _objectWithoutPropertiesLoose(props, _excluded);

  var _useContext = useContext(Context),
      prefix = _useContext.prefix,
      breakPoint = _useContext.breakPoint,
      sectionGap = _useContext.sectionGap;

  var footerCls = classNames(className, (_classNames = {}, _classNames[prefix + "page-footer"] = true, _classNames[prefix + "page-footer--fixed"] = fixed, _classNames[prefix + "page-footer--no-top-padding"] = noTopPadding, _classNames[prefix + "page-footer--dividing"] = divider, _classNames[prefix + "bg--" + mode] = mode, _classNames));

  var newStyle = _extends({}, isValidGap(sectionGap) ? {
    marginBottom: wrapUnit(sectionGap)
  } : null, style);

  var innerStyle = _extends({
    maxWidth: wrapUnit(breakPoint.maxContentWidth, 'px')
  }, containerStyle);

  return createElement(View, _extends({}, others, {
    className: footerCls,
    style: newStyle
  }), createElement(View, {
    className: prefix + "page-footer-inner " + containerClassName,
    style: innerStyle
  }, children, createElement(SafeArea, null)));
};

Footer.displayName = 'Footer';
Footer._typeMark = 'Footer';
Footer.defaultProps = {
  mode: 'surface',
  divider: false
};
export default Footer;
