// @alifd/mobile-layout/es/page/header.js
var _excluded = ["children", "divider", "className", "background", "mode", "fixed", "containerClassName", "containerStyle", "noBottomPadding", "style"];

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import { createElement, useContext } from 'rax';
import classNames from 'classnames';
import View from '../common/view';
import Context from '../common/context';
import { isValidGap, wrapUnit } from '../utils';

/**
 * 序章
 * @param props
 * @constructor
 */
var Header = function (props) {
  var _classNames;

  var children = props.children,
      _props$divider = props.divider,
      divider = _props$divider === void 0 ? false : _props$divider,
      className = props.className,
      background = props.background,
      mode = props.mode,
      fixed = props.fixed,
      containerClassName = props.containerClassName,
      containerStyle = props.containerStyle,
      noBottomPadding = props.noBottomPadding,
      style = props.style,
      others = _objectWithoutPropertiesLoose(props, _excluded);

  var _useContext = useContext(Context),
      prefix = _useContext.prefix,
      breakPoint = _useContext.breakPoint,
      sectionGap = _useContext.sectionGap;

  var headerCls = classNames(className, (_classNames = {}, _classNames[prefix + "page-header"] = true, _classNames[prefix + "page-header--fixed"] = fixed, _classNames[prefix + "page-header--no-bottom-padding"] = noBottomPadding, _classNames[prefix + "page-header--dividing"] = divider, _classNames[prefix + "bg--" + mode] = mode, _classNames));

  var newStyle = _extends({}, isValidGap(sectionGap) ? {
    marginBottom: wrapUnit(sectionGap)
  } : null, style);

  var innerStyle = _extends({
    maxWidth: wrapUnit(breakPoint.maxContentWidth, 'px')
  }, containerStyle);

  return createElement(View, _extends({}, others, {
    className: headerCls,
    style: newStyle
  }), createElement(View, {
    className: prefix + "page-header-inner " + containerClassName,
    style: innerStyle
  }, children));
};

Header.displayName = 'Header';
Header._typeMark = 'Header';
Header.defaultProps = {
  mode: 'surface',
  divider: false,
  noBottomPadding: false
};
export default Header;
