// @alifd/mobile-layout/es/page/page.js
var _excluded = ["className", "prefix", "mode", "responsive", "breakPoints", "children", "sectionGap", "blockGap", "gridGap", "noPadding", "onBreakPointChange"];

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import { Fragment } from "rax";
import { createElement, useEffect } from 'rax';
import classNames from 'classnames';
import Children from 'rax-children';
import isValidElement from 'rax-is-valid-element';
import View from '../common/view';
import { getMaxNumberOfColumns } from '../utils';
import Context from '../common/context';
import Header from './header';
import Content from './content';
import Footer from './footer';
import useWindowResize from '../utils/hooks/use-window-resize';
import { DEFAULT_BREAK_POINTS } from '../common/constant';

/**
 * 过滤出 content header footer
 * @param children
 */
function reformChildren(children) {
  var header = null;
  var footer = null;
  var content = null;
  var otherChildren = Children.map(children, function (child) {
    if (isValidElement(child)) {
      var _child$type;

      var tm = child == null ? void 0 : (_child$type = child.type) == null ? void 0 : _child$type._typeMark;

      if (tm === 'Header') {
        if (header) {
          // eslint-disable-next-line no-console
          console.error('[mobile-pro-layout]: <Page> should have only one <Header> child');
        }

        header = child;
        return null;
      } else if (tm === 'Footer') {
        if (footer) {
          // eslint-disable-next-line no-console
          console.error('[mobile-pro-layout]: <Page> should have only one <Footer> child');
        }

        footer = child;
        return null;
      } else if (tm === 'Content') {
        if (content) {
          // eslint-disable-next-line no-console
          console.error('[mobile-pro-layout]: <Page> should have only one <Content> child');
        }

        content = child;
        return null;
      }
    }

    return child;
  }).filter(function (child) {
    return !!child;
  });
  return createElement(Fragment, null, header, content || createElement(Content, null, otherChildren), footer);
}
/**
 * 页面
 * @param props
 * @constructor
 */


var Page = function (props) {
  var _classNames;

  var className = props.className,
      prefix = props.prefix,
      mode = props.mode,
      responsive = props.responsive,
      breakPoints = props.breakPoints,
      children = props.children,
      sectionGap = props.sectionGap,
      blockGap = props.blockGap,
      gridGap = props.gridGap,
      noPadding = props.noPadding,
      onBreakPointChange = props.onBreakPointChange,
      others = _objectWithoutPropertiesLoose(props, _excluded);

  var curBreakPoint = useWindowResize(breakPoints, onBreakPointChange, responsive); // 当前断点支持的最大列数

  var maxNumberOfColumns = getMaxNumberOfColumns(breakPoints);
  var pageCls = classNames(className, (_classNames = {}, _classNames[prefix + "page"] = true, _classNames[prefix + "bg--" + mode] = mode, _classNames[prefix + "page--col-" + curBreakPoint.numberOfColumns] = true, _classNames[prefix + "page--no-padding"] = noPadding, _classNames));
  useEffect(function () {
    if (onBreakPointChange) {
      // 页面加载后，直接触发一次断点回调
      onBreakPointChange(curBreakPoint, null, breakPoints);
    }
  }, []);
  return createElement(View, _extends({}, others, {
    className: pageCls
  }), createElement(Context.Provider, {
    value: {
      prefix: prefix,
      breakPoint: curBreakPoint,
      maxNumberOfColumns: maxNumberOfColumns,
      noPadding: noPadding,
      sectionGap: sectionGap,
      blockGap: blockGap,
      gridGap: gridGap
    }
  }, reformChildren(children)));
};

Page.displayName = 'Page';
Page._typeMark = 'Page';
Page.Header = Header;
Page.Content = Content;
Page.Footer = Footer;
Page.defaultProps = {
  prefix: 'mt-layout-',
  mode: 'lining',
  responsive: false,
  breakPoints: DEFAULT_BREAK_POINTS,
  noPadding: false
};
export default Page;
