// @alifd/mobile-layout/lib/local/col-grid.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _rax = require("rax");

var _raxChildren = _interopRequireDefault(require("rax-children"));

var _raxIsValidElement = _interopRequireDefault(require("rax-is-valid-element"));

var _utils = require("../utils");

var _grid = _interopRequireDefault(require("./grid"));

var _excluded = ["children", "width", "height", "style", "gap"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var ColGrid = function (props) {
  var children = props.children,
      width = props.width,
      height = props.height,
      style = props.style,
      gap = props.gap,
      others = _objectWithoutPropertiesLoose(props, _excluded);

  var columns = [];

  var newChildren = _raxChildren.default.map(children, function (child) {
    if (!child) {
      return null;
    }

    if ((0, _raxIsValidElement.default)(child)) {
      var _child$props = child.props,
          childStyle = _child$props.style,
          childHeight = _child$props.height,
          autoFit = _child$props.autoFit;
      var validHeight;

      if (childStyle != null && childStyle.height && (childStyle == null ? void 0 : childStyle.height) !== '') {
        validHeight = childStyle.height;
      } else if (childHeight && childStyle !== '') {
        validHeight = childHeight;
      }

      if (autoFit) {
        columns.push('fit-content(1%)');
      } else if (validHeight) {
        // notice: 如果是有明确的高度，那么使用高度作为自适应处理
        columns.push("minmax(" + (0, _utils.wrapUnit)(validHeight) + ", auto)"); // columns.push('auto');
        // columns.push(wrapUnit(validHeight));
      } else {
        columns.push('1fr');
      }
    } else {
      columns.push('auto');
    }

    return child;
  });

  var newStyle = _extends({
    gridTemplateRows: columns.map(function (c) {
      return (0, _utils.wrapUnit)(c, 'fr');
    }).join(' ')
  }, width && width > 0 ? {
    width: (0, _utils.wrapUnit)(width)
  } : null, height && height > 0 ? {
    height: (0, _utils.wrapUnit)(height)
  } : null, style);

  return (0, _rax.createElement)(_grid.default, _extends({}, others, {
    cols: 1,
    rowGap: gap,
    style: newStyle
  }), newChildren);
};

ColGrid.displayName = 'ColGrid';
var _default = ColGrid;
exports.default = _default;
