// @alifd/mobile-layout/lib/local/p.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _rax = require("rax");

var _classnames = _interopRequireDefault(require("classnames"));

var _raxChildren = _interopRequireDefault(require("rax-children"));

var _view = _interopRequireDefault(require("../common/view"));

var _context = _interopRequireDefault(require("../common/context"));

var _text = _interopRequireDefault(require("./text"));

var _utils = require("../utils");

var _useGuid = _interopRequireDefault(require("../utils/hooks/use-guid"));

var _excluded = ["type", "className", "beforeMargin", "afterMargin", "align", "verAlign", "spacing", "verMargin", "style", "children"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var getNewChildren = function (children, clsPrefix, type, pid) {
  // 如果无 children, 依然保持空行
  if (!children) {
    return (0, _rax.createElement)(_text.default, {
      className: clsPrefix + "-text"
    }, " ");
  }

  return _raxChildren.default.map(children, function (child, index) {
    // 包裹文本节点已处理各方向上间距
    if (typeof child === 'string') {
      return (0, _rax.createElement)(_text.default, {
        key: pid + "-" + index,
        className: clsPrefix + "-text",
        type: type
      }, child);
    }

    return child;
  });
};
/**
 * 段落布局，自动为段落内元素增加横向和垂直间距，并支持多种模式对齐
 */


var P = function (props) {
  var _classNames;

  var type = props.type,
      className = props.className,
      beforeMargin = props.beforeMargin,
      afterMargin = props.afterMargin,
      align = props.align,
      verAlign = props.verAlign,
      spacing = props.spacing,
      verMargin = props.verMargin,
      style = props.style,
      children = props.children,
      others = _objectWithoutPropertiesLoose(props, _excluded);

  var _useContext = (0, _rax.useContext)(_context.default),
      prefix = _useContext.prefix;

  var pid = (0, _useGuid.default)('pid');
  var clsPrefix = prefix + "p";
  var paragraphCls = (0, _classnames.default)(clsPrefix, className, (_classNames = {}, _classNames[clsPrefix + "-spacing"] = spacing !== false, _classNames[clsPrefix + "-align--" + align] = align, _classNames[clsPrefix + "-valign--" + verAlign] = verAlign, _classNames[clsPrefix + "-spacing--" + align] = spacing && align, _classNames[clsPrefix + "-spacing--" + spacing] = ['small', 'medium', 'large'].indexOf(spacing) > -1, _classNames[clsPrefix + "-margin"] = verMargin, _classNames[clsPrefix + "--" + type] = type, _classNames));

  var newStyle = _extends({
    marginTop: (0, _utils.wrapUnit)(beforeMargin),
    marginBottom: (0, _utils.wrapUnit)(afterMargin)
  }, style);

  return (0, _rax.createElement)(_view.default, _extends({}, others, {
    className: paragraphCls,
    style: newStyle
  }), getNewChildren(children, clsPrefix, type, pid));
};

P.displayName = 'P';
P._typeMark = 'P';
P.defaultProps = {
  spacing: 'medium',
  verMargin: true,
  verAlign: 'middle',
  beforeMargin: 0,
  afterMargin: 0
};
var _default = P;
exports.default = _default;
