// @alifd/mobile-layout/lib/local/row-flex.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _rax = require("rax");

var _raxChildren = _interopRequireDefault(require("rax-children"));

var _classnames = _interopRequireDefault(require("classnames"));

var _raxIsValidElement = _interopRequireDefault(require("rax-is-valid-element"));

var _raxCloneElement = _interopRequireDefault(require("rax-clone-element"));

var _context = _interopRequireDefault(require("../common/context"));

var _constant = require("../common/constant");

var _utils = require("../utils");

var _excluded = ["children", "className", "verAlign", "style", "width", "height", "gap"],
    _excluded2 = ["width"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var RowFlex = function (props) {
  var children = props.children,
      className = props.className,
      verAlign = props.verAlign,
      style = props.style,
      width = props.width,
      height = props.height,
      gapProp = props.gap,
      others = _objectWithoutPropertiesLoose(props, _excluded);

  var _useContext = (0, _rax.useContext)(_context.default),
      prefix = _useContext.prefix,
      gridGap = _useContext.gridGap;

  var gap = (0, _utils.getGapVal)(gridGap, gapProp);

  var newChildren = _raxChildren.default.map(children, function (child, index) {
    if ((0, _raxIsValidElement.default)(child)) {
      var _child$props = child == null ? void 0 : child.props,
          childWidth = _child$props.width,
          autoFit = _child$props.autoFit,
          childStyle = _child$props.style;

      var _ref = childStyle || {},
          childStyleWidth = _ref.width,
          otherChildStyle = _objectWithoutPropertiesLoose(_ref, _excluded2);

      var flex;
      var validWidth;

      if (childWidth && childWidth !== '') {
        validWidth = childWidth;
      } else if (childStyleWidth && childStyleWidth !== '') {
        validWidth = childStyle.width;
      }

      if (autoFit) {
        flex = '0 0 auto';
      } else if (validWidth) {
        flex = "0 0 " + (0, _utils.wrapUnit)(validWidth);
      } else {
        flex = '1 1 0';
      }

      return (0, _raxCloneElement.default)(child, _extends({}, child == null ? void 0 : child.props, {
        key: "layout-row-flex-" + index,
        style: _extends({
          flex: flex
        }, otherChildStyle)
      }));
    }

    return child;
  });

  var newStyle = _extends({
    alignItems: _constant.VER_ALIGN_ALIAS_MAP[verAlign] || verAlign
  }, (0, _utils.isValidGap)(gap) && gap !== 'auto' ? {
    gap: (0, _utils.wrapUnit)(gap)
  } : null, width && width > 0 ? {
    width: (0, _utils.wrapUnit)(width)
  } : null, height && height > 0 ? {
    height: (0, _utils.wrapUnit)(height)
  } : null, style);

  return (0, _rax.createElement)("div", _extends({}, others, {
    className: (0, _classnames.default)(className, prefix + "row-flex"),
    style: newStyle
  }), newChildren);
};

RowFlex.displayName = 'RowFlex';
var _default = RowFlex;
exports.default = _default;
