// @uni/system-info/es/web/getInfo.js
import { formatSystemInfo, formatSystemInfoAsync } from "../common";
import { CONTAINER_NAME } from "../_utils/constant";

function getPlatform() {
  var _navigator = navigator,
      platform = _navigator.platform;

  if (platform) {
    if (/(iphone|ipod|ipad)/.test(navigator.userAgent.toLowerCase())) return 'iOS';
    if (/android/.test(navigator.userAgent.toLowerCase())) return 'Android';
    return platform;
  }

  return '';
}

var getInfoSync = formatSystemInfo(function () {
  return {
    pixelRatio: window.devicePixelRatio,
    screenWidth: window.document.documentElement.clientWidth,
    screenHeight: window.document.documentElement.clientHeight,
    windowWidth: window.innerWidth,
    windowHeight: window.innerHeight,
    language: navigator.language,
    version: navigator.appVersion,
    platform: getPlatform() || ''
  };
});
var getInfo = formatSystemInfoAsync(function (options) {
  var success = options.success,
      fail = options.fail,
      complete = options.complete;

  try {
    var res = getInfoSync();
    success && success(res);
    complete && complete(res);
  } catch (e) {
    fail && fail(e);
    complete && complete(e);
  }
}, CONTAINER_NAME.WEB);
export default getInfo;
