// @alifd/mobile-layout/lib/page/block.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _babelRuntimeJsxPlus = require("babel-runtime-jsx-plus");

var _rax = require("rax");

var _classnames = _interopRequireDefault(require("classnames"));

var _context = _interopRequireDefault(require("../common/context"));

var _cell = _interopRequireDefault(require("./cell"));

var _row = _interopRequireDefault(require("../local/row"));

var _p = _interopRequireDefault(require("../local/p"));

var _col = _interopRequireDefault(require("../local/col"));

var _utils = require("../utils");

var _excluded = ["className", "title", "titleAlign", "extra", "noPadding", "mode", "bordered", "span", "contentStyle", "contentClassName", "align", "verAlign", "divider", "children"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * 区块,默认 flex 布局
 * @param props
 * @constructor
 */
var Block = function (props) {
  var _classNames, _classNames2, _classNames3;

  var className = props.className,
      title = props.title,
      titleAlign = props.titleAlign,
      extra = props.extra,
      noPaddingProp = props.noPadding,
      mode = props.mode,
      bordered = props.bordered,
      spanProp = props.span,
      contentStyle = props.contentStyle,
      contentClassName = props.contentClassName,
      align = props.align,
      verAlign = props.verAlign,
      divider = props.divider,
      children = props.children,
      others = _objectWithoutPropertiesLoose(props, _excluded);

  var _useContext = (0, _rax.useContext)(_context.default),
      prefix = _useContext.prefix,
      maxNumberOfColumns = _useContext.maxNumberOfColumns;

  var isTransparent = mode === 'transparent';
  var noPadding = !isTransparent ? noPaddingProp : true;
  var clsPrefix = prefix + "block"; // 当有 title 或 extra 节点时

  var hasHead = !isTransparent && (title || extra);
  var span = spanProp; // 保证 span 无效时, 撑满整行 section

  if (!span || span > maxNumberOfColumns || span <= 0) {
    span = maxNumberOfColumns;
  }

  var blockCls = (0, _classnames.default)(className, (_classNames = {}, _classNames[prefix + "block"] = true, _classNames[prefix + "bg--" + mode] = mode && !isTransparent, _classNames[prefix + "block--span-" + span] = span > 0, _classNames[prefix + "block--headless"] = !hasHead, _classNames[prefix + "block--no-padding"] = noPadding, _classNames[prefix + "block--bordered"] = bordered, _classNames));
  var headCls = (0, _classnames.default)((_classNames2 = {}, _classNames2[clsPrefix + "-head"] = true, _classNames2[clsPrefix + "--divided"] = divider && hasHead, _classNames2));
  var blockContentCls = (0, _classnames.default)(contentClassName, (_classNames3 = {}, _classNames3[prefix + "block-content"] = true, _classNames3[prefix + "block-content--no-padding"] = noPadding, _classNames3));

  if (hasHead) {
    return (0, _rax.createElement)(_col.default, _extends({}, others, {
      className: blockCls
    }), (0, _rax.createElement)(_row.default, {
      autoFit: true,
      verAlign: "middle",
      className: headCls
    }, (0, _babelRuntimeJsxPlus.createCondition)([[function () {
      return title;
    }, function () {
      return (0, _rax.createElement)(_cell.default, {
        className: prefix + "block-title",
        align: titleAlign
      }, (0, _utils.isString)(title) ? (0, _rax.createElement)(_p.default, {
        type: "body1",
        align: "left"
      }, title) : title);
    }]]), (0, _babelRuntimeJsxPlus.createCondition)([[function () {
      return extra;
    }, function () {
      return (0, _rax.createElement)(_cell.default, {
        autoFit: true,
        align: "right",
        className: prefix + "block-extra"
      }, (0, _utils.isString)(extra) ? (0, _rax.createElement)(_p.default, {
        type: "body2",
        align: "right"
      }, extra) : extra);
    }]])), (0, _rax.createElement)(_cell.default, {
      align: align,
      verAlign: verAlign,
      className: blockContentCls,
      style: contentStyle
    }, children));
  }

  return (0, _rax.createElement)(_cell.default, _extends({}, others, {
    className: blockCls
  }), children);
};

Block.displayName = 'Block';
Block._typeMark = 'Block';
Block.defaultProps = {
  mode: 'surface',
  noPadding: false
};
var _default = Block;
exports.default = _default;
