// @alifd/mobile-layout/lib/page/footer.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _rax = require("rax");

var _classnames = _interopRequireDefault(require("classnames"));

var _view = _interopRequireDefault(require("../common/view"));

var _context = _interopRequireDefault(require("../common/context"));

var _utils = require("../utils");

var _safeArea = _interopRequireDefault(require("../common/safe-area"));

var _excluded = ["children", "className", "divider", "noTopPadding", "fixed", "mode", "containerClassName", "containerStyle", "style"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * 终章
 * @param props
 * @constructor
 */
var Footer = function (props) {
  var _classNames;

  var children = props.children,
      className = props.className,
      divider = props.divider,
      noTopPadding = props.noTopPadding,
      fixed = props.fixed,
      mode = props.mode,
      containerClassName = props.containerClassName,
      containerStyle = props.containerStyle,
      style = props.style,
      others = _objectWithoutPropertiesLoose(props, _excluded);

  var _useContext = (0, _rax.useContext)(_context.default),
      prefix = _useContext.prefix,
      breakPoint = _useContext.breakPoint,
      sectionGap = _useContext.sectionGap;

  var footerCls = (0, _classnames.default)(className, (_classNames = {}, _classNames[prefix + "page-footer"] = true, _classNames[prefix + "page-footer--fixed"] = fixed, _classNames[prefix + "page-footer--no-top-padding"] = noTopPadding, _classNames[prefix + "page-footer--dividing"] = divider, _classNames[prefix + "bg--" + mode] = mode, _classNames));

  var newStyle = _extends({}, (0, _utils.isValidGap)(sectionGap) ? {
    marginBottom: (0, _utils.wrapUnit)(sectionGap)
  } : null, style);

  var innerStyle = _extends({
    maxWidth: (0, _utils.wrapUnit)(breakPoint.maxContentWidth, 'px')
  }, containerStyle);

  return (0, _rax.createElement)(_view.default, _extends({}, others, {
    className: footerCls,
    style: newStyle
  }), (0, _rax.createElement)(_view.default, {
    className: prefix + "page-footer-inner " + containerClassName,
    style: innerStyle
  }, children, (0, _rax.createElement)(_safeArea.default, null)));
};

Footer.displayName = 'Footer';
Footer._typeMark = 'Footer';
Footer.defaultProps = {
  mode: 'surface',
  divider: false
};
var _default = Footer;
exports.default = _default;
