// @alifd/mobile-layout/lib/page/header.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _rax = require("rax");

var _classnames = _interopRequireDefault(require("classnames"));

var _view = _interopRequireDefault(require("../common/view"));

var _context = _interopRequireDefault(require("../common/context"));

var _utils = require("../utils");

var _excluded = ["children", "divider", "className", "background", "mode", "fixed", "containerClassName", "containerStyle", "noBottomPadding", "style"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

/**
 * 序章
 * @param props
 * @constructor
 */
var Header = function (props) {
  var _classNames;

  var children = props.children,
      _props$divider = props.divider,
      divider = _props$divider === void 0 ? false : _props$divider,
      className = props.className,
      background = props.background,
      mode = props.mode,
      fixed = props.fixed,
      containerClassName = props.containerClassName,
      containerStyle = props.containerStyle,
      noBottomPadding = props.noBottomPadding,
      style = props.style,
      others = _objectWithoutPropertiesLoose(props, _excluded);

  var _useContext = (0, _rax.useContext)(_context.default),
      prefix = _useContext.prefix,
      breakPoint = _useContext.breakPoint,
      sectionGap = _useContext.sectionGap;

  var headerCls = (0, _classnames.default)(className, (_classNames = {}, _classNames[prefix + "page-header"] = true, _classNames[prefix + "page-header--fixed"] = fixed, _classNames[prefix + "page-header--no-bottom-padding"] = noBottomPadding, _classNames[prefix + "page-header--dividing"] = divider, _classNames[prefix + "bg--" + mode] = mode, _classNames));

  var newStyle = _extends({}, (0, _utils.isValidGap)(sectionGap) ? {
    marginBottom: (0, _utils.wrapUnit)(sectionGap)
  } : null, style);

  var innerStyle = _extends({
    maxWidth: (0, _utils.wrapUnit)(breakPoint.maxContentWidth, 'px')
  }, containerStyle);

  return (0, _rax.createElement)(_view.default, _extends({}, others, {
    className: headerCls,
    style: newStyle
  }), (0, _rax.createElement)(_view.default, {
    className: prefix + "page-header-inner " + containerClassName,
    style: innerStyle
  }, children));
};

Header.displayName = 'Header';
Header._typeMark = 'Header';
Header.defaultProps = {
  mode: 'surface',
  divider: false,
  noBottomPadding: false
};
var _default = Header;
exports.default = _default;
