// @alifd/mobile-layout/lib/page/section.js
"use strict";

exports.__esModule = true;
exports.default = void 0;

var _babelRuntimeJsxPlus = require("babel-runtime-jsx-plus");

var _rax = require("rax");

var _classnames = _interopRequireDefault(require("classnames"));

var _raxChildren = _interopRequireDefault(require("rax-children"));

var _raxCloneElement = _interopRequireDefault(require("rax-clone-element"));

var _context = _interopRequireDefault(require("../common/context"));

var _view = _interopRequireDefault(require("../common/view"));

var _utils = require("../utils");

var _row = _interopRequireDefault(require("../local/row"));

var _col = _interopRequireDefault(require("../local/col"));

var _p = _interopRequireDefault(require("../local/p"));

var _block = _interopRequireDefault(require("./block"));

var _cell = _interopRequireDefault(require("./cell"));

var _useGuid = _interopRequireDefault(require("../utils/hooks/use-guid"));

var _excluded = ["className", "children", "title", "noPadding", "extra", "gap", "style"];

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

/**
 * 获取计算后的子节点
 * @param children
 * @param numberOfColumns
 * @param maxNumberOfColumns
 * @param sid
 */
function getValidChildren(children, numberOfColumns, maxNumberOfColumns, sid) {
  var newChildren = wrapBlock(children, maxNumberOfColumns);
  return calBlockSpan(newChildren, numberOfColumns, sid);
}
/**
 * 为非 Block 节点，包裹 Block 子元素
 * @param children
 * @param maxNumberOfColumns
 * @return RaxNode[]
 */


function wrapBlock(children, maxNumberOfColumns) {
  var tmp = [];
  var ret = [];

  var validChildList = _raxChildren.default.toArray(children).filter(function (child) {
    return !!child;
  });

  validChildList.forEach(function (child, index) {
    if (child.type === _block.default) {
      ret.push(child);

      if (tmp.length > 0) {
        ret.push((0, _rax.createElement)(_block.default, {
          span: maxNumberOfColumns
        }, [].concat(tmp)));
        tmp = [];
      }
    } else {
      tmp.push(child);
    }

    if (index === validChildList.length - 1 && tmp.length > 0) {
      ret.push((0, _rax.createElement)(_block.default, {
        span: maxNumberOfColumns
      }, [].concat(tmp)));
      tmp = [];
    }
  });
  return ret;
}
/**
 * 为 Block 自动调整列宽
 * @param children
 * @param numberOfColumns
 * @param sid
 */


function calBlockSpan(children, numberOfColumns, sid) {
  var ret = [];
  var stack = [];
  var counter = 0;
  var len = children.length;

  for (var i = 0; i < len; i++) {
    var child = children[i];
    var span = child.props.span;
    var gid = sid + "-" + numberOfColumns + "-" + i;

    if (span < numberOfColumns) {
      if (span + counter <= numberOfColumns) {
        stack.push(child);
        counter += span;
      } else {
        ret = [].concat(ret, adjustColWidth(stack, counter, numberOfColumns, gid));
        stack = [child];
        counter = span;
      }
    } else {
      if (stack.length > 0) {
        ret = [].concat(ret, adjustColWidth(stack, counter, numberOfColumns, gid));
        stack = [];
        counter = 0;
      }

      ret = [].concat(ret, adjustColWidth([child], numberOfColumns, numberOfColumns, gid + "-r2"));
    }

    if (i === len - 1 && stack.length > 0) {
      ret = [].concat(ret, adjustColWidth(stack, counter, numberOfColumns, gid + "-last"));
    }
  }

  return ret;
}
/**
 * 重算列宽
 * @param blockNodes
 * @param totalSpan
 * @param maxColNum
 * @param sid
 */


function adjustColWidth(blockNodes, totalSpan, maxColNum, sid) {
  return blockNodes.map(function (item, index) {
    var span = item.props.span;
    return (0, _raxCloneElement.default)(item, _extends({}, item.props, {
      id: sid + "-" + index,
      key: sid + "-" + index,
      span: Math.round(span / totalSpan * maxColNum)
    }));
  });
}
/**
 * 章节
 * @param props
 */


var Section = function (props) {
  var _classNames, _classNames2;

  var _useContext = (0, _rax.useContext)(_context.default),
      prefix = _useContext.prefix,
      blockGapContext = _useContext.blockGap,
      _useContext$breakPoin = _useContext.breakPoint,
      numberOfColumns = _useContext$breakPoin.numberOfColumns,
      maxContentWidth = _useContext$breakPoin.maxContentWidth,
      maxNumberOfColumns = _useContext.maxNumberOfColumns;

  var className = props.className,
      children = props.children,
      title = props.title,
      noPadding = props.noPadding,
      extra = props.extra,
      blockGapProp = props.gap,
      style = props.style,
      others = _objectWithoutPropertiesLoose(props, _excluded);

  var clsPrefix = prefix + "section";
  var bid = (0, _useGuid.default)('sc');
  var gap = (0, _utils.getGapVal)(blockGapContext, blockGapProp);
  var sectionCls = (0, _classnames.default)(clsPrefix, className);
  var blockWrapperCls = (0, _classnames.default)((_classNames = {}, _classNames[clsPrefix + "-block-wrapper"] = true, _classNames[prefix + "block--auto"] = gap === 'auto', _classNames));
  var ctnWithHeadCls = (0, _classnames.default)(clsPrefix + "-inner", clsPrefix + "-inner--with-head", (_classNames2 = {}, _classNames2[prefix + "block--auto"] = gap === 'auto', _classNames2[clsPrefix + "-no-padding"] = noPadding, _classNames2));
  var ctnWithoutHeadCls = (0, _classnames.default)(clsPrefix + "-inner", clsPrefix + "-inner-without-head");

  var blockStyle = _extends({}, (0, _utils.isValidGap)(gap) && gap !== 'auto' ? {
    gridColumnGap: (0, _utils.wrapUnit)(gap),
    gridRowGap: (0, _utils.wrapUnit)(gap)
  } : null, style);

  var blockWrapperStyle = _extends({}, (0, _utils.isValidGap)(gap) && gap !== 'auto' ? {
    gridColumnGap: (0, _utils.wrapUnit)(gap),
    gridRowGap: (0, _utils.wrapUnit)(gap)
  } : null);

  var innerStyle = {
    maxWidth: (0, _utils.wrapUnit)(maxContentWidth, 'px')
  };

  if (title || extra) {
    return (0, _rax.createElement)(_view.default, _extends({}, others, {
      className: (0, _classnames.default)(clsPrefix, className)
    }), (0, _rax.createElement)(_view.default, {
      className: ctnWithHeadCls,
      style: innerStyle
    }, (0, _rax.createElement)(_col.default, {
      align: "stretch"
    }, (0, _rax.createElement)(_row.default, {
      autoFit: true,
      verAlign: "middle",
      className: clsPrefix + "-head"
    }, (0, _babelRuntimeJsxPlus.createCondition)([[function () {
      return title;
    }, function () {
      return (0, _rax.createElement)(_cell.default, {
        className: clsPrefix + "-title"
      }, (0, _utils.isString)(title) ? (0, _rax.createElement)(_p.default, {
        align: "left",
        type: "body1"
      }, title) : title);
    }]]), (0, _babelRuntimeJsxPlus.createCondition)([[function () {
      return extra;
    }, function () {
      return (0, _rax.createElement)(_cell.default, {
        autoFit: true,
        align: "right",
        className: clsPrefix + "-extra"
      }, (0, _utils.isString)(extra) ? (0, _rax.createElement)(_p.default, {
        align: "right",
        type: "body2"
      }, extra) : extra);
    }]])), (0, _rax.createElement)(_cell.default, null, (0, _rax.createElement)("div", {
      className: blockWrapperCls,
      style: blockWrapperStyle
    }, getValidChildren(children, numberOfColumns, maxNumberOfColumns, bid))))));
  }

  return (0, _rax.createElement)(_view.default, _extends({}, others, {
    className: sectionCls,
    style: blockStyle
  }), (0, _rax.createElement)(_view.default, {
    className: (0, _classnames.default)(ctnWithoutHeadCls, blockWrapperCls),
    style: _extends({}, blockWrapperStyle, innerStyle)
  }, getValidChildren(children, numberOfColumns, maxNumberOfColumns, bid)));
};

Section.displayName = 'Section';
Section._typeMark = 'Section';
var _default = Section;
exports.default = _default;
